src/lib.rs

```diff
@@ -119,6 +123,38 @@ mod tests {
 		assert_eq!("foo", read_b);
 	}
 
+	#[cfg(target_os = "macos")]
+	#[test]
+	fn test_swap_dirs() {
+		use self::libloading::os::unix::{Library, Symbol};
+
+		lazy_static! {
+			/// `renamex_np` is available only on macos >= 10.12
+			static ref RENAMEX_NP: Option<Symbol<unsafe extern fn (oldpath: *const libc::c_char, newpath: *const libc::c_char, flags: libc::c_uint) -> libc::c_int>> = unsafe {
+				let lib = Library::this();
+				lib.get(b"renamex_np").ok()
+			};
+		}
+		let dir_a = TempDir::new("a").unwrap();
+		let dir_b = TempDir::new("b").unwrap();
+		let path_a = dir_a.path().join("file");
+		let path_b = dir_b.path().join("file");
+		write_to_file(&path_a, "foo");
+		write_to_file(&path_b, "bar");
+		// `renamex_np` available test if `recursive swapping` works
+		if let Some(ref _renamex_np) = &*RENAMEX_NP {
```

Comment:
Unfortunately existence of this function does not guarantee that swap works. It works only for APFS volumes

Reply:
haha, after this I obviously love OSX but this is the thing we need I guess https://developer.apple.com/documentation/foundation/nsurlvolumesupportsswaprenamingkey

Reply:
this function is just a NSFoundation wrapper for `getattrlist`.

>           RENAME_SWAP   On file systems that support it (see getattrlist(2)
                         VOL_CAP_INT_RENAME_SWAP), it will cause the source
                         and target to be atomically swapped.  Source and tar-
                         get need not be of the same type, i.e. it is possible
                         to swap a file with a directory.  EINVAL is returned
                         in case of bitwise-inclusive OR with RENAME_EXCL.

link: http://www.manpagez.com/man/2/renamex_np/osx-10.12.3.php